compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1725,7 +1725,12 @@ impl<'a> Parser<'a> {
                 err.help("see https://doc.rust-lang.org/book/ch05-03-method-syntax.html for more information");
                 err
             } else {
-                self.expected_ident_found()
+                let mut err = self.expected_ident_found();
+                if ident.name == kw::Let {
```

Comment:
Ideally we would do N=1 lookahead to check if the next element is an identifier, and then do `self.bump();` (to consume the `let`) and then get that ident, *emit the error* and return `Ok(ident)` (so that the parser can properly recover and continue creating the `struct` with the expected field name, and even emit errors if the mistyped field isn't included elsewhere). In all other cases, we keep the current behavior (returning `err`) and let the other parser recovery kick in.

Reply:
We might want to extend the existing test to include something like

```rust
struct S {
    let s: (),
    y: (),
}
fn main() {
    let _ = S { // should complain about missing `y`
        s: (),
    };
    let _ = S {
        y: (),  // should complain about missing `s`
    };
    let _ = S {
        let s: (),  // should complain about missing `y` and the spurious `let`
    };
    let _ = S {
        let y: (), // should complain about missing `s` and the spurious `let`
    };
    let _ = S {
        let s = (),  // should complain about missing `y`, the `:` to `=` typo and the spurious `let`
    };
    let _ = S {
        let y = (), // should complain about missing `s`, the `:` to `=` typo and the spurious `let`
    };
}
```

Keep in mind that the later four recoveries are not included in this PR (but I would like to have it in the future).